### ISSUE TITLE
Proposal: require branches to be up to date before merging

### ISSUE THREAD

**Issue by HonkingGoose** [progit/progit2#1488]
**Which version of the book is affected?**

Problem is with the current pull request merge process on GitHub.

**Describe the bug:**

Because of the current repository branch protection settings for `master`, GitHub allows maintainers to merge in work that is not fully tested with the current state of the `master` branch. It also puts the responsibility for fixing the broken work on the maintainers shoulders, instead of on the contributors shoulders. "If you break the build it's yours to fix it."

**Steps to reproduce:**

1. Make a pull request.
2. Mainline branch has newer commits that are not merged into the pull branch.
3. Notice that if there are no merge-conflicts, the maintainer can merge the now stale pull request with the passing tests on the stale pull request. The maintainer thinks: "Nice work and the tests are passing, let's merge!"
4. Maintainer merges stale pull request.
5. This forces the tests to run again on the mainline branch.
6. Tests fail because of a logical error between the branches, which Git will not catch.
6.1 Or tests fail because the pull request is not fully compatible with the changes made on mainline.
7. Be annoyed: "Why let me merge stuff that is broken?"
8. Maintainer of project is now on the hook to fix the broken stuff.

**Expected behavior:**

1. When a pull request is out-of-date with `master`, merging is blocked by GitHub.
2. The contributor/maintainer updates the branch by merging in the current mainline branch.
3. The tests are automatically run again, to ensure mainline and the pull branch are actually functional together.
3.1 If the tests fail, the contributor will need to fix their work so that the tests are passing again.
4. Only when the tests pass on a fully updated pull branch is the work merged.

**Additional context:**

This change increases the integration frequency, which is a good thing.
By integrating more often, the exact commit where things break is easier to find.
This also provides some protection against egregious logical errors, if the tests cover the logical error:

1. Pull is dependent on a file being present on mainline.
2. Mainline deletes file needed by the pull.
3. This causes no merge conflicts, so GitHub will let you merge it.
4. Tests runs on master, now the logical error is caught, or another user flags the brokenness.

**Benefits of change:**

- Prevents some nasty surprises ("It worked before, why won't it work now???")
- Puts burden of compatibility with the pull request author.
- Ensures the mainline branch has a greater chance of being fully functional at all times.
- Allows contributors/maintainers to catch logical errors before they get on the mainline branch.
- Keeps changes small, if the branch is kept up to date, each change is small, and it's easier to find where things break.

**Drawbacks of change:**

- This change forces a rerun of all the tests, which costs time. The maintainer must wait for the new tests to pass before merging.
- Total TravisCI running time is increased, which might increase the bills the project needs to pay.
- Puts burden of compatibility with the pull request author.

**How to make the proposed change as repository administrator:**
Steps:
1. Login to Github.
2. Go to the **Settings** tab for this repo.
3. In the left menu, click **Branches**.
4. In the  "Branch protection rules" section, select the `master` branch.
5. Under "Protect matching branches", select Require status checks to pass before merging. (This checkbox is already ticked).
6. There's a setting called: "Require branches to be up to date before merging", tick this checkbox.
7. Optionally: force the administrators to follow the same process as the outside contributors by ticking the box "Include administrators".

GitHub's Official documentation and helpful screenshots can be found here: https://docs.github.com/en/github/administering-a-repository/enabling-required-status-checks

**Comment by ben** [comment]
Done!

**Comment by HonkingGoose** [comment]
## Apology:

I'm very sorry @ben, I should not nudge you into things without giving you the full picture. I completely forgot to mention a huge drawback: making the change causes a lot of merges on the pull branch. And that without a change in the pull request merge process, those merges will clutter up the `git log --all --oneline --graph --all` overview.

I was aware of this problem, but it fully slipped from my mind when I was drafting this proposal. I've ran into this issue multiple times on working with other repositories... So I really should have warned you.

I apologize for not mentioning this before nudging you into this change. 😞
I will do better on the any new proposals, and ensure the drawbacks are properly covered.

Now to make amends, I have listed the full problem, and ways to remedy it below.

## Description of problem:

Each time a contributor/administrator clicks the "Update branch" button on the GitHub pull interface, it creates a new standard merge. When you merge the pull into the main branch, the `git log --all --decorate --oneline --graph` history will be a mess of merges. This complicates reading the merge history and figuring out what was changed. I tried using the `git log --all --decorate --oneline -graph --no-merges` option, but that hides the merge commit from the git log output. So that is not helpful when you want to revert the merge commit.

## Ways to remedy the problem:

- Make no GitHub configuration changes and ask the pull request owner to `git rebase -i upstream/main` their work after the final review, and let them clobber over the branch by force pushing after the final review. But this will not scale when `main` moves ahead with newer commits under the pull request authors feet. So this leads to some kind of weird whack-a-mole, and causes a lot of back and forth discussion with the pull request owner.

- Turn of the default merge and allow [commit-squash-merges](https://docs.github.com/en/github/administering-a-repository/configuring-commit-squashing-for-pull-requests) from within the GitHub pull interface. Administrators and those with direct write access can still do a normal merge from the local workstation and push it. This means that all commits will be squashed into a single commit, and then make their way into the `main` branch. The benefit here is that the single commit you're left with is easy to revert (a bit like the way you can revert a bad merge by reverting the merge commit).

- Turn of the default merge and allow [rebasing merge](https://docs.github.com/en/github/administering-a-repository/about-merge-methods-on-github#rebasing-and-merging-your-commits). Administrators and those with direct write access can still do a normal merge from the local workstation and push it. This has both a benefit and drawback: when there is a single commit, it's applied cleanly to the `main` branch. If there are multiple commits, each of those will make it's way into `main` but the changes made individually might not be able to be cleanly backed out of with `git revert`. So this requires each commit to be tested and functional. The Travis CI only guarantees that the last pushed commit works, commits in between are not tested. So this might lead to breakage when trying to revert a single commit.

- Or [tick the Require linear history](https://docs.github.com/en/github/administering-a-repository/requiring-a-linear-commit-history#about-enforcement-of-linear-commit-history) option in the `master` branch protection rules. This would only allow squash/rebase merges, and turns of the standard merge. Administrators and those with direct write access can still do a normal merge from the local workstation and push it.

- If none of the above appeals: revert the **Require branches to be up to date before merging** option in the branch protection tab.

**Comment by ben** [comment]
Yeah, I'm starting to feel the pain of this configuration change. I can't open 5 PRs, review them, and hit merge if they look okay. After the first one, I need to click a button and wait for CI to run.

Honestly I don't think we've ever had a situation where this rule would have saved us from a headache. Changes in one chapter generally won't break content in another, or the build in general. I'm going to roll back that config change.

**Comment by HonkingGoose** [comment]
@ben Just for your information:

GitHub now allows the repository maintainer to allow "auto-merge". This enables a new workflow:

1. Review PR.
2. Merge `main` branch into PR branch with button.
3. Tell GitHub to automerge if test pass with button.
4. Do something else with the free time.
5. GitHub merges PR once all test pass.
6. CI creates release
7. Done.

Now for branches that **must** be up-to-date with main, it's still not 100% idiot proof, and you'll stilll need to re-merge `main` when things change. So it's not as easy as it could be.

To learn more about automerge, you can read: https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request

---

I understand that this is still more work when reviewing and merging 5 PRs "the old way". But it might still make things faster when reviewing a single PR. 😉 

I'm not proposing that we turn on "keep branches up to date before merging" again, as that was clearly too much of a hassle. 😉

**Comment by ben** [comment]
Just turned on auto-merge. I've got it turned on for a few other projects too, it's pretty nice.
